content/backend/graphql-scala/2-preparing-first-query.md

```diff
@@ -0,0 +1,377 @@
+
+### Define GraphQLSchema
+
+<Instruction>
+
+Create GraphQLSchema object:
+
+```scala
+import com.howtographql.scala.sangria.models.Link
+import sangria.schema.{Field, ListType, ObjectType}
+// #
+import sangria.schema._
+import sangria.macros.derive._
+
+object GraphQLSchema {
+
+  // 1
+  implicit val LinkType = deriveObjectType[Unit, Link]()
+
+  // 2
+  val QueryType = ObjectType(
+    "Query",
+    fields[MyContext, Unit](
+      Field("allLinks", ListType(LinkType), resolve = c => c.ctx.dao.allLinks)
+    )
+  )
+
+  // 3
+  val SchemaDefinition = Schema(QueryType)
+}
+```
+
+</Instruction>
+
+Sangria cannot reuse case classes defied in our domain, it needs it's own object called of type `ObjectType`. On the other hand it no need to be 1-1 mapping between our Sangria and DB models. This abstraction allows us to freely hide, add or aggregate fields.
+* **1** is a definition of ObjectType for our `Link` class. `deriveObjectType` is a macro what will built an Object for us. But we can do it manually. For example `deriveObjectType[Unit, Link]()` is the same to
+
+```scala
+val LinkType = ObjectType[Unit, Link](
+     "Link",
+     fields[Unit, Link](
+       Field("id", IntType, resolve = _.value.id),
+       Field("url", StringType, resolve = _.value.url),
+       Field("description", StringType, resolve = _.value.description)
+     )
+   )
+```
+
+When you want to use macro-way to define objects don't forget to import `sangria.macros.derive._`
+
+* **2** `val QueryType` is a top level object of our schema. Probably it also could be defined by macro but I decided to make it manually. As you can see, the top level object has name `Query` and it (along with nested object) will be available to see in the graphiql console what we will include farther in this chapter. In `fields` definition I've added only one `Field` at this moment.
+
+```scala
+
+  Field("allLinks", ListType(LinkType), resolve = c => c.ctx.dao.allLinks)
+
+```
+
+Means that we want to expose a field with name "allLinks", response is a list of `LinkType` objects. At the end we have to provide a function wich will resolve a result. In this case it's a `DAO`'s function `allLinks`. We have to implement it now.
+
+<Instruction>
+
+Add a function to the `DAO` object:
+
+```scala
+
+def allLinks = db.run(Links.result)
+
+```
+
+</Instruction>
+
+
+### GraphiQL console
+
+[Graphiql](https://github.com/graphql/graphiql) makes able to run queries against our server from the browser. Let's imlement it now.
+
+<Instruction>
+
+All we need to do, is to copy `graphiql.html` file into the project folder. You can use original file, but I have a one that is slightly modified which works out-of-th-box so you can use it aswell.
+
+Download a graphiql.html file from [link](https://github.com/howtographql/sangria/blob/1_getting_started/src/main/resources/graphiql.html) 
+
+and save into the `/src/main/resources` directory
```

Comment:
One of the "busy-work" that g8 template could do for us

Reply:
:+1: moved file to giter8 template